### DISCUSSION TITLE
Uncaught ReferenceError: WaveSurfer is not defined

### DISCUSSION THREAD

**Discussion by RLangridge** [katspaugh/wavesurfer.js#2942] [Q&A]
Hi there, I'm trying to run WaveSurfer in an Electron project. I'm using the following:
Node v18.16
Electron v25.2
WaveSurfer v7.0.0-beta.11

When I try to use the annotation plugin (I've installed wave surfer via yarn), I get the following error when I import:

import RegionsPlugin from 'wavesurfer.js/dist/plugins/regions.js';
Uncaught ReferenceError: WaveSurfer is not defined
This comes from the minified js from the regions plugin. Anyone have any idea what's happening here?

I get the same error when using nupkg

**Comment by katspaugh** [comment]
It sounds like you forgot to import WaveSurfer?

**Comment by RLangridge** [comment]
Sorry; I should have said that I had already imported WaveSurfer:
This is in React btw:

import WaveSurfer from 'wavesurfer.js';
import RegionsPlugin from 'wavesurfer.js/dist/plugins/regions.js';
... 

const ws = WaveSurfer.create({
      container: document.getElementById(`wavesurfer-${index}`),
      waveColor: 'rgb(200, 0, 200)',
      progressColor: 'rgb(100, 0, 100)',
      url: `${saveLocation}${path.sep}SessionAudio${path.sep}Downloads${path.sep}${file.filename}`,
    });

    const wsRegions = ws.registerPlugin(RegionsPlugin.create());

Below is the error:
`Uncaught ReferenceError: WaveSurfer is not defined
    at Object.<anonymous> (G:\Programming\Work\Redenlab\RedenlabAnalyzeManager\node_modules\wavesurfer.js\dist\plugins\regions.min.cjs:1:351)
    at Object.<anonymous> (G:\Programming\Work\Redenlab\RedenlabAnalyzeManager\node_modules\wavesurfer.js\dist\plugins\regions.min.cjs:2:3)
    at Module._compile (node:internal/modules/cjs/loader:1269:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1324:10)
    at Module.load (node:internal/modules/cjs/loader:1124:32)
    at Module._load (node:internal/modules/cjs/loader:965:12)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at o._load (node:electron/js2c/renderer_init:2:3109)
    at Module.require (node:internal/modules/cjs/loader:1148:19)
    at require (node:internal/modules/cjs/helpers:110:18)`

**Reply by katspaugh** [reply]
What bundler are you using? Wepback/rollup?

**Comment by RLangridge** [comment]
Sorry I'm going to seem like a fool here but I'm using babel and that's about it. electron-builder is doing the build to create the electron app.

**Reply by katspaugh** [reply]
It's all good! Can you post your babel config?

**Reply by RLangridge** [reply]
{ "presets": ["es2015", "stage-0", "react"], "plugins": ["transform-decorators-legacy", "transform-runtime"], }
That's the babel config. This is a bit of an old project so I'm running an older version of babel (v6.26)

**Reply by katspaugh** [reply]
It seems that it's trying to import a UMD/CommonJS module but cannot see WaveSurfer in the current scope for some reason.
I've no idea how to fix this tbh, other than maybe inserting wavesurfer and the regions plugin as `<script>` tags in your HTML.

I have an example with React and Babel here and it works fine: https://wavesurfer-js.org/examples/#react.js

**Reply by RLangridge** [reply]
Ok I'll give that a try and report back with the results. Thanks for your time.

**Reply by RLangridge** [reply]
Hi. Unfortunately after spending a good chunk of the day trying to get this running, I'm still hitting issues. So I've done the following since I was last here:
- Updated babel to latest version. babel.config.json now looks like this:

```
{
  "presets": ["@babel/preset-env", "@babel/preset-react"],
  "plugins": ["@babel/plugin-transform-modules-commonjs"],
  "sourceType": "module"
}
```
- Tried using the unpkg (gives me this error:
`Uncaught SyntaxError: Cannot use import statement outside a module (at wavesurfer.js@beta:1:1)`
- Tried importing via scripts:
```
    <script src="https://unpkg.com/wavesurfer.js@beta"></script>
    <script src="https://unpkg.com/wavesurfer.js@beta/dist/plugins/regions.js"></script>
```
This would also render the same issues as above (import statement)
- Attempted to add the module argument to the scripts:

```
    <script module src="https://unpkg.com/wavesurfer.js@beta"></script>
    <script module src="https://unpkg.com/wavesurfer.js@beta/dist/plugins/regions.js"></script>
```

This unfortunately doesn't seem to give me a way to retrieve WaveSurfer (I was hoping I could just grab it from window, that doesn't seem to be the case).

Any ideas as to what else I can try here? I was really hoping to use WaveSurfer with the regions plugin.

Thanks for your time

**Reply by katspaugh** [reply]
Please check the readme, it specifies the correct way to include a script tag.

**Reply by RLangridge** [reply]
My apologies it's been a long day I hadn't realised I needed to change it over to min.cjs. Unfortunately this has just run me into another issue claiming that I can't run the script due to the following error:

`Refused to execute script from 'https://unpkg.com/wavesurfer.js@7.0.0-beta.11/dist/plugins/regions.min.cjs' because its MIME type ('text/plain') is not executable, and strict MIME type checking is enabled.`

At this point, would it be better if I pull down the repo and run WaveSurfer locally?

**Reply by katspaugh** [reply]
No problem, the readme is admittedly not very clear about the ES modules vs CommonJS distinction. I assumed devs already know about this.

That's a bummer that the script MIME type isn't recognized. Looks like unpkg doesn't serve cjs files with the right content type. You might be able to add a type to the tag like this:

```html
<script type="text/javascript" src="https://unpkg.com/wavesurfer.js@beta/dist/wavesurfer.min.cjs"></script>
<script type="text/javascript" src="https://unpkg.com/wavesurfer.js@beta/dist/plugins/regions.min.cjs"></script>
```

You could also trying going back to your original approach and import the scripts like this:

```js
import WaveSurfer from 'wavesurfer.js/dist/wavesurfer.min.cjs'
import RegionsPlugin from 'wavesurfer.js/dist/plugins/regions.min.cjs'
```

**Reply by RLangridge** [reply]
Thanks for your help and time with this. Unfortunately neither approach is too content. 

With approach 1 I'm still getting the same where it's refusing to run due to the MIME type not being recognised. I've tried passing through the type before as "application/json" just to try my luck. That didn't crash but it didn't succeed in adding WaveSurfer to window. 

With approach 2 I've received the following error:
`Uncaught Error: Package subpath './dist/wavesurfer.min.cjs' is not defined by "exports" in G:\app\node_modules\wavesurfer.js\package.json`

**Reply by katspaugh** [reply]
That error is actually good, it means Babel/Node.js are able to correctly recognize which file to import during the build as per the package.json.

So your original imports
```js
import WaveSurfer from 'wavesurfer.js'
import RegionsPlugin from 'wavesurfer.js/dist/plugins/regions.js'
```
are correctly routed to the corresponding cjs files, but then it fails to find the WaveSurfer global var (although it shouldn't even be looking for it).

I'm starting to think it's a problem with how I configured Webpack to build CommonJS modules for plugins. It shouldn't export them to the global WaveSurfer namespace if it's imported as a module.

Needs more research on my side, please hold on. :)

**Reply by RLangridge** [reply]
Great thank you for looking into this :)

**Reply by katspaugh** [reply]
I've just released beta.12 with a fixed export, could you try again? Using your original import syntax.

**Reply by RLangridge** [reply]
That's fixed it! I now no longer get that issue and I can use the Regions plugin as per normal. Thank you so much for looking into this.

**Reply by katspaugh** [reply]
Awesome, thank you!

**Reply by sicklyman** [reply]
I also have the same issue. All was working last week on my wordpress install at https://jgsongs.com but today none of the posts display wavesurfer. The console says 
![image](https://github.com/katspaugh/wavesurfer.js/assets/4551431/32cc88c5-d7e1-4309-84a1-c8eb33676a73)

and the code I'm using in each post is as follows:

`    <script src="https://unpkg.com/wavesurfer.js"></script>
    <script>
const playButton = document.querySelector("#playButton")
const playButtonIcon = document.querySelector("#playButtonIcon")
const waveform = document.querySelector("#waveform")
// --------------------------------------------------------- //
/**
 * Initialize Wavesurfer
 * @returns a new Wavesurfer instance
 */
const initializeWavesurfer = () => {
  return WaveSurfer.create({
    container: "#waveform",
    responsive: true,
    cursorColor: '#fff',
    waveColor: '#B17435',
    progressColor: '#676767',
    barWidth: 2,
    barRadius: 2,
    cursorWidth: 2,
    height: 70,
    barGap: 1

  })
}
// --------------------------------------------------------- //
// Functions
/**
 * Toggle play button
 */
const togglePlay = () => {
  wavesurfer.playPause()
  const isPlaying = wavesurfer.isPlaying()
  if (isPlaying) {
    playButtonIcon.src = "https://jgsongs.com/wp-content/uploads/2023/01/pauseGray.png"
  } else {
    playButtonIcon.src = "https://jgsongs.com/wp-content/uploads/2023/01/playGray.png"
  }
}

// Create a new instance and load the wavesurfer
const wavesurfer = initializeWavesurfer()
wavesurfer.load("https://jgsongs.com/wp-content/uploads/2023/03/Psalm-45.mp3")

// Javascript Event listeners
playButton.addEventListener("click", togglePlay)

// Resets the play button icon after audio ends
wavesurfer.on("finish", () => {
  playButtonIcon.src = "https://jgsongs.com/wp-content/uploads/2023/01/playGray.png"
})
</script>`

Any help appreciated.

**Reply by katspaugh** [reply]
Sorry about that!
You should use this script tag:
```html
<script src="https://unpkg.com/wavesurfer.js/dist/wavesurfer.min.js"></script>
```

**Reply by sicklyman** [reply]
brilliant! That worked. 

Any idea why it worked fine until just this week?

**Reply by katspaugh** [reply]
Because I released a new version yesterday, v7, that has a different default export. I'm changing it back here: #3008.

Edit: `<script src="https://unpkg.com/wavesurfer.js"></script>` should also work again now.

**Comment by sicklyman** [comment]
right... this makes me feel uneasy. From now on, I guess I should use your updated script tag, right? I'm just worried about future updates breaking the plugin. What's the best way to keep up to date with your new releases?

**Reply by sicklyman** [reply]
wonderful plugin by the way. I should have said that at the start. My apologies!

**Reply by katspaugh** [reply]
No worries, I totally get the concern.
It’s generally a good idea to pin the version of your CDN scripts, otherwise it will always fetch the latest one which is a bit like a Trojan horse. With the version pinned you can be deliberate about your updates.
So I apologize that the readme in this repo was going against this advice. I’ve updated it now to include a version.
I.e.
```html
<script src="https://unpkg.com/wavesurfer.js@7/dist/wavesurfer.min.js"></script>
```
Hope this makes sense and sorry for the breakage today. Thanks for letting me know btw, your report helped others. Glad we fixed this relatively early.
